### PR TITLE
Add Screenshot Feature

### LINKED ISSUE #2143: [Feature Request] Add a button to take a screenshot on the UI
#### Overview:
Add button to take a screenshot on the UI, this way we don´t have to rely on a 3rd party software to take screenshots. A hotkey to this purpose could also be added.

#### Smaller Details: 
The screenshot could be saved on \publish\screenshots\

#### Nature of Request:  
- Addition
 
### Why would this feature be useful?
To not rely on 3rd party software

### PR DESCRIPTION
**Pull request by emmauss**
Adds taking screen captures of games at their internal resolution. So you can now take your 8K screenshots even if you don't have a display that high. 
It can be used from the Actions menu or with the hotkey, F8.

Fixes #2143 .

### CODE REVIEW

File: Ryujinx/Configuration/ConfigurationState.cs [hunk 1]
```diff
@@ -534,7 +534,8 @@ public void LoadDefault()
             Hid.EnableKeyboard.Value               = false;
             Hid.Hotkeys.Value = new KeyboardHotkeys
             {
-                ToggleVsync = Key.Tab
+                ToggleVsync = Key.Tab,
+                ScreenShot = Key.F8
```
Comment (AcK77) [inline Ryujinx/Configuration/ConfigurationState.cs:538]: `ScreenShot` -> `Screenshot`

File: Ryujinx/Configuration/ConfigurationState.cs [hunk 2]
```diff
@@ -828,6 +829,19 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
...
             }
 
+            if (configurationFileFormat.Version < 27)
+            {
+                Common.Logging.Logger.Warning?.Print(LogClass.Application, $"Outdated configuration version {configurationFileFormat.Version}, migrating to version 27.");
+
+                configurationFileFormat.Hotkeys = new KeyboardHotkeys
+                {
+                    ToggleVsync = Key.Tab,
+                    ScreenShot = Key.F8
```
Comment (AcK77) [inline Ryujinx/Configuration/ConfigurationState.cs:839]: `ScreenShot` -> `Screenshot`

File: Ryujinx.Common/Configuration/Hid/KeyboardHotkeys.cs [hunk 3]
```diff
@@ -3,5 +3,6 @@
     public struct KeyboardHotkeys
     {
         public Key ToggleVsync { get; set; }
+        public Key ScreenShot { get; set; }
```
Comment (AcK77) [inline Ryujinx.Common/Configuration/Hid/KeyboardHotkeys.cs:6]: `ScreenShot` -> `Screenshot`

File: Ryujinx.Graphics.GAL/IRenderer.cs [hunk 4]
```diff
@@ -44,5 +46,7 @@ public interface IRenderer : IDisposable
         void WaitSync(ulong id);
 
         void Initialize(GraphicsDebugLevel logLevel);
+
+        void ScreenShot();
```
Comment (AcK77) [inline Ryujinx.Graphics.GAL/IRenderer.cs:50]: `ScreenShot` -> `Screenshot`

File: Ryujinx.Graphics.GAL/ScreenCaptureImageInfo.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        {
+            Width  = width;
+            Height = height;
+            IsBgra = isBgra;
+            Data   = data;
+            FlipX  = flipX;
+            FlipY  = flipY;
+        }
+
+        public int Width { get; }
```
Comment (AcK77) [inline Ryujinx.Graphics.GAL/ScreenCaptureImageInfo.cs:15]: If you align assignment above then align var names and `{ get; }` (some spaces are missing)

File: Ryujinx.Graphics.OpenGL/Renderer.cs [hunk 6]
```diff
@@ -195,5 +197,15 @@ public void WaitSync(ulong id)
         {
             _sync.Wait(id);
         }
+
+        public void ScreenShot()
```
Comment (AcK77) [inline Ryujinx.Graphics.OpenGL/Renderer.cs:201]: `ScreenShot` -> `Screenshot`

File: Ryujinx.Graphics.OpenGL/Window.cs [hunk 7]
```diff
@@ -16,6 +16,8 @@ class Window : IWindow, IDisposable
 
         internal BackgroundContextWorker BackgroundContext { get; private set; }
 
+        internal bool ScreenCaptureRequested{ get; set; }
```
Comment (AcK77) [inline Ryujinx.Graphics.OpenGL/Window.cs:19]:
```suggestion
        internal bool ScreenCaptureRequested { get; set; }
```

File: Ryujinx.Graphics.OpenGL/Window.cs [hunk 8]
```diff
@@ -106,6 +108,13 @@ private void CopyTextureToFrameBufferRGB(int drawFramebuffer, int readFramebuffe
             int dstY0 = crop.FlipY ? dstPaddingY : _height - dstPaddingY;
             int dstY1 = crop.FlipY ? _height - dstPaddingY : dstPaddingY;
 
+            if(ScreenCaptureRequested)
```
Comment (AcK77) [inline Ryujinx.Graphics.OpenGL/Window.cs:111]:
```suggestion
            if (ScreenCaptureRequested)
```

File: Ryujinx.Graphics.OpenGL/Window.cs [hunk 9]
```diff
@@ -159,6 +168,14 @@ public void InitializeBackgroundContext(IOpenGLContext baseContext)
             BackgroundContext = new BackgroundContextWorker(baseContext);
         }
 
+        public void CaptureFrame(int x, int y, int width, int height, bool isBgra, bool flipX, bool flipY)
+        {
+            long size = Math.Abs(4 * width * height);
+            byte[] bitmap = new byte[size];
+            GL.ReadPixels(x, y, width, height, isBgra ? PixelFormat.Bgra : PixelFormat.Rgba, PixelType.UnsignedByte, bitmap);
```
Comment (AcK77) [inline Ryujinx.Graphics.OpenGL/Window.cs:175]: Add a new empty line before and after

File: Ryujinx/Ui/MainWindow.cs [hunk 10]
```diff
@@ -1344,6 +1350,14 @@ private void Scan_Amiibo(object sender, EventArgs args)
             }
         }
 
+        private void Take_Screenshot(object sender, EventArgs args)
+        {
+            if(_emulationContext!=null && RendererWidget != null)
```
Comment (AcK77) [inline Ryujinx/Ui/MainWindow.cs:1355]:
```suggestion
            if (_emulationContext != null && RendererWidget != null)
```

File: Ryujinx/Ui/MainWindow.glade [hunk 11]
```diff
@@ -1,92 +1,89 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- Generated with glade 3.22.1 -->
+<!-- Generated with glade 3.38.2 -->
```
Comment (AcK77) [inline Ryujinx/Ui/MainWindow.glade:2]: Why all those changes? Does this cause no issues?

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 12]
```diff
@@ -31,6 +38,8 @@ public abstract class RendererWidgetBase : DrawingArea
         public Switch Device { get; private set; }
         public IRenderer Renderer { get; private set; }
 
+        public bool ScreenshotRequested{ get; set; }
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:41]:
```suggestion
        public bool ScreenshotRequested { get; set; }
```

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 13]
```diff
@@ -299,9 +308,54 @@ public void Initialize(Switch device)
             Renderer = Device.Gpu.Renderer;
             Renderer?.Window.SetSize(_windowWidth, _windowHeight);
 
+            if(Renderer != null)
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:311]:
```suggestion
            if (Renderer != null)
```

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 14]
```diff
@@ -299,9 +308,54 @@ public void Initialize(Switch device)
...
 
+        private unsafe void Renderer_ScreenCaptured(object sender, ScreenCaptureImageInfo e)
+        {
+            if (e.Data.Length > 0)
+            {
+                Task.Run(() =>
+                {
+                    lock (this)
+                    {
+                        var currentTime  = DateTime.Now;
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:327]: If you align assignments you should align var names too

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 15]
```diff
@@ -299,9 +308,54 @@ public void Initialize(Switch device)
...
+                        string filename  = $"ryujinx_capture_{currentTime.Year}-{currentTime.Month:D2}-{currentTime.Day:D2}_{currentTime.Hour:D2}-{currentTime.Minute:D2}-{currentTime.Second:D2}.png";
+                        string directory = System.IO.Path.Combine(System.Environment.GetFolderPath(System.Environment.SpecialFolder.MyPictures), "Ryujinx");
+                        string path      = System.IO.Path.Combine(directory, filename);
+
+                        Directory.CreateDirectory(directory);
+
+                        Image image = e.IsBgra ? Image.LoadPixelData<Bgra32>(e.Data, e.Width, e.Height)
+                                    : Image.LoadPixelData<Rgba32>(e.Data, e.Width, e.Height);
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:335]: Maybe align the : with the ?

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 16]
```diff
@@ -299,9 +308,54 @@ public void Initialize(Switch device)
...
+
+                        if (e.FlipX)
+                        {
+                            image.Mutate(x => x.Flip(FlipMode.Horizontal));
+                        }
+                        if (e.FlipY)
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:341]: Add a new empty line before

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 17]
```diff
@@ -299,9 +308,54 @@ public void Initialize(Switch device)
...
+                        }
+
+                        image.SaveAsPng(path, new PngEncoder()
+                        {
+                            ColorType = PngColorType.Rgb
+                        });
+
+                        image.Dispose();
+
+                        Logger.Notice.Print(LogClass.Application, $"ScreenShot saved to {path}", "Screenshot");
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:353]: `ScreenShot` -> `Screenshot`

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 18]
```diff
@@ -573,7 +635,8 @@ private bool UpdateFrame()
         private enum KeyboardHotkeyState
         {
             None,
-            ToggleVSync
+            ToggleVSync,
+            ScreenShot
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:639]: `ScreenShot` -> `Screenshot`

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 19]
```diff
@@ -584,6 +647,11 @@ private KeyboardHotkeyState GetHotkeyState()
             {
                 state |= KeyboardHotkeyState.ToggleVSync;
             }
+            
+            if (_keyboardInterface.IsPressed((Key)ConfigurationState.Instance.Hid.Hotkeys.Value.ScreenShot))
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:651]: `ScreenShot` -> `Screenshot`

File: Ryujinx/_schema.json [hunk 20]
```diff
@@ -1443,14 +1443,22 @@
       "type": "object",
       "title": "Hotkey Controls",
       "required": [
-        "toggle_vsync"
+        "toggle_vsync",
+        "screen_shot"
```
Comment (AcK77) [inline Ryujinx/_schema.json:1447]:
```suggestion
        "screenshot"
```

File: Ryujinx/_schema.json [hunk 21]
```diff
@@ -1443,14 +1443,22 @@
...
+        
       ],
       "properties": {
         "toggle_vsync": {
           "$id": "#/properties/hotkeys/properties/toggle_vsync",
           "$ref": "#/definitions/key",
           "title": "Toggle VSync",
           "default": "Tab"
+        },
+        "screen_shot": {
```
Comment (AcK77) [inline Ryujinx/_schema.json:1457]:
```suggestion
        "screenshot": {
```

File: Ryujinx/_schema.json [hunk 22]
```diff
@@ -1443,14 +1443,22 @@
...
+          "$id": "#/properties/hotkeys/properties/screen_shot",
```
Comment (AcK77) [inline Ryujinx/_schema.json:1458]:
```suggestion
          "$id": "#/properties/hotkeys/properties/screenshot",
```

File: Ryujinx/_schema.json [hunk 23]
```diff
@@ -1443,14 +1443,22 @@
...
+          "$ref": "#/definitions/key",
+          "title": "ScreenShot",
```
Comment (AcK77) [inline Ryujinx/_schema.json:1460]:
```suggestion
          "title": "Screenshot",
```

File: Ryujinx/Config.json [hunk 24]
```diff
@@ -56,7 +56,8 @@
   "start_fullscreen": false,
   "enable_keyboard": false,
   "hotkeys": {
-    "toggle_vsync": "Tab"
+    "toggle_vsync": "Tab",
+    "screen_shot": "F8"
```
Comment (AcK77) [inline Ryujinx/Config.json:60]:
```suggestion
    "screenshot": "F8"
```

**Review by AcK77 — CHANGES_REQUESTED** [review]

File: Ryujinx/Ui/MainWindow.glade (hunk 11 above)
Comment (emmauss) [reply Ryujinx/Ui/MainWindow.glade:2]: nope. it's metadata glade adds. it doesn't affect the ui builder.

**Comment by emmauss** [comment]
Fixed the names

**Comment by GamerzHell9137** [comment]
Any more information about it? Like where the SS are being saved?

**Comment by GamerzHell9137** [comment]
For people who want to test the PR, the folder where the screenshots are being saved is Windows Picture folder.

**Comment by Benedani** [comment]
While I don't have concerns about the Pictures folder on Linux (where it exists), I think the path should be made configurable in the settings.

File: Ryujinx/Config.json [hunk 25]
```diff
@@ -56,7 +56,8 @@
   "start_fullscreen": false,
   "enable_keyboard": false,
   "hotkeys": {
-    "toggle_vsync": "Tab"
+    "toggle_vsync": "Tab",
+    "screenshot": "F8"
```
Comment (Thog) [inline Ryujinx/Config.json:60]: Also need a configuration version change

File: Ryujinx/_schema.json [hunk 26]
```diff
@@ -1443,14 +1443,22 @@
       "type": "object",
       "title": "Hotkey Controls",
       "required": [
-        "toggle_vsync"
+        "toggle_vsync",
+        "screenshot"
+
```
Comment (Thog) [inline Ryujinx/_schema.json:1448]: Remove empty line

File: Ryujinx/_schema.json [hunk 27]
```diff
@@ -1443,14 +1443,22 @@
...
           "$id": "#/properties/hotkeys/properties/toggle_vsync",
           "$ref": "#/definitions/key",
           "title": "Toggle VSync",
           "default": "Tab"
+        },
+        "screenshot": {
+          "$id": "#/properties/hotkeys/properties/screenshot",
+          "$ref": "#/definitions/key",
+          "title": "Screenshot",
+          "default": "F9"
```
Comment (Thog) [inline Ryujinx/_schema.json:1461]:
```suggestion
          "default": "F8"
```

**Review by Thog — CHANGES_REQUESTED** [review]
Some small issues otherwise lgtm

File: Ryujinx/Config.json (hunk 25 above)
Comment (emmauss) [reply Ryujinx/Config.json:60]: I've already bumped the version from 26 to 27

**Review by Thog — APPROVED** [review]
code lgtm, haven't got much time to test if this cause any regressions in term of performances (cc @gdkchan & @EmulationFanatic)

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 28]
```diff
@@ -272,10 +281,56 @@ public void Initialize(Switch device)
...
 
+        private unsafe void Renderer_ScreenCaptured(object sender, ScreenCaptureImageInfo e)
+        {
+            if (e.Data.Length > 0)
+            {
+                Task.Run(() =>
+                {
+                    lock (this)
+                    {
+                        var    currentTime  = DateTime.Now;
```
Comment (AcK77) [inline Ryujinx/Ui/RendererWidgetBase.cs:301]:
```suggestion
                        var    currentTime = DateTime.Now;
```
and same for other vars

File: Ryujinx/Ui/MainWindow.glade (hunk 11 above)
Comment (AcK77) [reply Ryujinx/Ui/MainWindow.glade:2]: I mean, it's not specifically this line, it's about the whole file. (Like `can_focus` to `can-focus` as example)

**Review by AcK77 — APPROVED** [review]

File: Ryujinx/Ui/MainWindow.glade [hunk 29]
```diff
@@ -292,8 +289,8 @@
                       <object class="GtkMenuItem" id="_simulateWakeUpMessage">
                         <property name="visible">True</property>
                         <property name="can_focus">False</property>
-                        <property name="tooltip_text" translatable="yes">Simulate a Wake-up Message</property>
-                        <property name="label" translatable="yes">Simulate Wake-up Message</property>
+                        <property name="tooltip_text" translatable="yes">Simulate a Wake_up Message</property>
```
Comment (riperiperi) [inline Ryujinx/Ui/MainWindow.glade:292]: Is this intentional? It looks like you meant to just change `-` to `_` in variable names.

**Review by riperiperi — COMMENTED** [review]
Nice feature! Just got one question with the changes to the glade file.

File: Ryujinx/Ui/MainWindow.glade (hunk 29 above)
Comment (emmauss) [reply Ryujinx/Ui/MainWindow.glade]: No. I've reverted it.

**Comment by emmauss** [comment]
rebased and ready to merge

**Comment by MelonSpeedruns** [comment]
Looks good to me, WHERE'S THE MERGE???? >:(

File: Ryujinx/Configuration/ConfigurationState.cs [hunk 30]
```diff
@@ -845,6 +846,19 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
                 configurationFileUpdated = true;
             }
 
+            if (configurationFileFormat.Version < 28)
+            {
+                Common.Logging.Logger.Warning?.Print(LogClass.Application, $"Outdated configuration version {configurationFileFormat.Version}, migrating to version 27.");
```
Comment (riperiperi) [inline Ryujinx/Configuration/ConfigurationState.cs:851]: Should be migrating to version 28

**Comment by GamerzHell9137** [comment]
Other than the rebasing thing everything works great. Gj!

File: Ryujinx/Configuration/ConfigurationState.cs (hunk 30 above)
Comment (emmauss) [reply Ryujinx/Configuration/ConfigurationState.cs]: fixed

**Review by riperiperi — APPROVED** [review]

**Pull request merged**